### PR TITLE
Improved password hashing using new specialized functions

### PR DESCRIPTION
**Pull request by lode**
PHP has specialized functions for password hashing, meant to simplify the interface but also to be more secure, partly because there's less room for implementation errors.

See also http://php.net/manual/en/book.password.php

Note I didn't check the database storage. This method will output something like `$2y$11$nsqRkXbim6GsNd5c976qFuYVhHh3vGNfwz8zvulg3UwudQThEqowK` and be 60 chars length with the current hashing algorithm defaults. Best is to use a `varchar(255)` or alike.

### CODE REVIEW

File: src/domains/Session/SessionService.php [hunk 1]
```diff
@@ -72,21 +76,20 @@ public function currentUser()
...
-            'salt' => bin2hex(random_bytes(32)),
-            'iterations' => 50000
-        );
+        // hash
+        $sHashedPassword = password_hash($sPassword, self::PASSWORD_HASH_ALGORITM, self::PASSWORD_HASH_OPTIONS);
 
-        // encrypt
-        $encryptedPassword->hash = hash_pbkdf2('sha256', $sPassword, $encryptedPassword->salt, $encryptedPassword->iterations, 64);
+        // validate
+        if (empty($sHashedPassword)) {
```
Comment (tstrijdhorst) [inline src/domains/Session/SessionService.php]: Check for explicit false:

https://secure.php.net/manual/en/function.password-hash.php

File: src/domains/Session/SessionService.php [hunk 2]
```diff
@@ -18,6 +18,10 @@
  */
 class SessionService
 {
+    const PASSWORD_HASH_ALGORITM = PASSWORD_BCRYPT;
```
Comment (tstrijdhorst) [inline src/domains/Session/SessionService.php]: Not sure if this is >PHP7, else use private const.

File: src/domains/Session/SessionService.php (hunk 1 above)
Comment (lode) [reply src/domains/Session/SessionService.php]: I explicitly didn't, as it is a fail-safe. If something incorrect comes out you'll want to catch it.

File: src/domains/Session/SessionService.php (hunk 2 above)
Comment (lode) [reply src/domains/Session/SessionService.php]: What, assigning a constant with a constant? That works on 5.6.

Comment (tstrijdhorst) [reply src/domains/Session/SessionService.php]: No I mean, I'm not sure if Mimoto has a >PHP7 requirement. If so you can use the new `private const` feature 😛

File: src/domains/Session/SessionService.php (hunk 1 above)
Comment (tstrijdhorst) [reply src/domains/Session/SessionService.php]: I get your point, but I don't think there is much merit to it. According to the interface it either returns a hashed string or it returns false. Unless you want to guard yourself against failures of the PHP Core (but then I'd regex the whole hash instead of just checking empty), this will never match anything else than false.

Now it makes it look like it can actually return an empty string, which is confusing imho.

File: src/domains/Session/SessionService.php (hunk 2 above)
Comment (lode) [reply src/domains/Session/SessionService.php]: Ah sorry, misread that part. There's at least no php requirement in the composer.json, so I assume 7 is not enforced.

File: src/domains/Session/SessionService.php (hunk 1 above)
Comment (lode) [reply src/domains/Session/SessionService.php]: Ha. I did this based on experience. Didn't know where from, until I found the compatibility package https://github.com/ircmaxell/password_compat again:

> It is very important that you should check the return value of `password_hash` prior to storing it, because `false` or `null` may be returned if it encountered an error.

I remember an earlier version also told to test against empty string or even 60 chars length. I got used to an empty check to prevent the worst.

Anyway, this also mainly depends on the coding standards in Mimoto. Whether `empty()` is to be avoided.

File: src/domains/Session/SessionService.php (hunk 2 above)
Comment (tstrijdhorst) [reply src/domains/Session/SessionService.php]: Yeah I saw that as well so I added this ticket:
https://github.com/SebastianKersten/Mimoto/issues/15

Comment (tstrijdhorst) [reply src/domains/Session/SessionService.php]: Ah, minimum php version is 5.6 so this is closed.
